src/Containers/Microsoft.NET.Build.Containers/Tasks/CreateNewImage.cs

```diff
@@ -325,4 +343,109 @@ public void Dispose()
+    /// <param name="LocalPullType">If the base image came from a local store of some kind, what kind of store was it?</param>
+    /// <param name="RemotePushType">If the new image is being pushed to a remote registry, what kind of registry is it?</param>
+    /// <param name="LocalPushType">If the new image is being stored in a local store of some kind, what kind of store is it?</param>
+    private record class PublishTelemetryContext(RegistryType? RemotePullType, LocalStorageType? LocalPullType, RegistryType? RemotePushType, LocalStorageType? LocalPushType);
+    private enum RegistryType { Azure, AWS, Google, GitHub, DockerHub, MCR, Other }
+    private enum LocalStorageType { Docker, Podman, Tarball }
+
+    private class Telemetry(Microsoft.Build.Utilities.TaskLoggingHelper Log, PublishTelemetryContext context)
+    {
+        private IDictionary<string, string?> ContextProperties() => new Dictionary<string, string?>
+            {
+                { nameof(context.RemotePullType), context.RemotePullType?.ToString() },
+                { nameof(context.LocalPullType), context.LocalPullType?.ToString() },
+                { nameof(context.RemotePushType), context.RemotePushType?.ToString() },
+                { nameof(context.LocalPushType), context.LocalPushType?.ToString() }
+            };
+
+        public void LogPublishSuccess()
+        {
+            Log.LogTelemetry("sdk/container/publish/success", ContextProperties());
+        }
+
+        public void LogUnknownRepository()
+        {
+            var props = ContextProperties();
+            props.Add("error", "unknown_repository");
+            Log.LogTelemetry("sdk/container/publish/error", props);
+        }
+
+        public void LogCredentialFailure(SourceImageReference _)
+        {
+            var props = ContextProperties();
+            props.Add("error", "credential_failure");
+            props.Add("direction", "pull");
+            Log.LogTelemetry("sdk/container/publish/error", props);
+        }
+
+        public void LogCredentialFailure(DestinationImageReference d)
+        {
+            var props = ContextProperties();
+            props.Add("error", "credential_failure");
+            props.Add("direction", "push");
+            Log.LogTelemetry("sdk/container/publish/error", props);
+        }
+
+        public void LogRidMismatch(string desiredRid, string[] availableRids)
+        {
+            var props = ContextProperties();
+            props.Add("error", "rid_mismatch");
+            props.Add("target_rid", desiredRid);
+            props.Add("available_rids", string.Join(",", availableRids));
+            Log.LogTelemetry("sdk/container/publish/error", props);
+        }
+
+        public void LogMissingLocalBinary()
+        {
+            var props = ContextProperties();
+            props.Add("error", "missing_binary");
+            Log.LogTelemetry("sdk/container/publish/error", props);
+        }
+
+        public void LogLocalLoadError()
```

Comment:
I dont think we have logic for unknown or unsupported media types for the images, nor for timeouts when uploading new image layers. Are those already included earlier or did we decide against them for now?